src/hardware/sblaster.cpp

```diff
@@ -1973,38 +1974,72 @@ static uint8_t CTMIXER_Read() {
 	return ret;
 }
 
+// Called by DSP_WRITE_STATUS to check if the write buffer is at capacity.
+static bool write_buffer_at_capacity()
+{
+	// Is the DSP in an abnormal state and therefore we should consider the
+	// buffer at capacity and unable to receive data?
+	if (sb.dsp.state != DSP_S_NORMAL) {
+		return true;
+	}
+	// Report the buffer as having some room every 8th call, as the buffer
+	// will have run down by some amount after sequential calls.
+	if ((++sb.dsp.write_status_counter % 8) == 0) {
+		return false;
+	}
+	// If DMA isn't running then the buffer's definitely not at capacity.
+	if (sb.dma.mode == DSP_DMA_NONE) {
+		return false;
+	}
+	// Finally, the DMA buffer is considered full until it's able to accept
+	// a full (64 KB) write; which is once we've hit the minimum threshold.
+	return (sb.dma.left > sb.dma.min);
```

Comment:
Are these new checks and conditions based on someone's reverse engineering work, or have you tried to do what's logical here @kcgen? If the latter, it's probably wrong as the behaviour of most hardware can be hardly derived from first principles... 😅 The whole SB behaviour is a minefield too; the TL;DR is only trust actual measurements, the doco is lying, etc.

I'm not saying I'm outright _against_ these extra checks, but try not to introduce too many ad-hoc constraints based on logic alone, that's my suggestion. I'd remove anything that's not absolutely essential, even if that means the buffer gets sometime ovewritten, etc. We just don't know how hardware works until we measure it.

I'm also confused about that modulo 8 thing (hack?) Do we know where that's coming from? Why not 11th call? 😅 Or why not report the actual reality? So, lots of questions...

I'm just cautious because when I screwed up a few things in my VGA refactor (single bits set wrong in very specific cases), it takes many months and a lot of testing for problems to surface (e.g. the Kyrandia scrolling bug). So all I'm saying, I highly suggest to be extremely conservative with these improvements 😄

Reply:
> I'm also confused about that modulo 8 thing (hack?) Do we know where that's coming from? Why not 11th call? 😅 Or why not report the actual reality? So, lots of questions...

Yes - I think it was a reasonable heuristic or perhaps someone measured it from their actual card. Unfortunately I'm not sure; but it wasn't perfect because games like SimLife simply give up if they're told they can't send data despite knowing the card is totally idle (probably 11 would be worse :laughing: )

We can see some of this in PaybackTime 2's pascal code: it does these door knocks 1000 times and then skips.

> re these new checks and conditions based on someone's reverse engineering work, or have you tried to do what's logical here @kcgen?

I created this based on logging what these games were doing (just a bunch of LOGs in SB) combined with reading that about what this check routine is meant to represent.

86Box and DOSBox-X both calculate the playback rate of the current DMA transfer and then steadily draining down that time until it nears completion.

In this implementation, fortunately we've already got the `dma.left` and `dma.min` values to see how close we are to the end of the transfer.

One of the nuances is that games can change the playback rate as well as switch from mono to stereo; so the drain down rates vary: and indeed, 86Box does this extra bookkeeping.

In our case, these rate changes are already taken care of by the existing code, and it just happens that this threshold (`dma.left` vs `dma.min`) will just happen sooner if the rates are faster.

Scali had a through discussion about Sound Blaster seamless playback using all the different DMA methods (autoinit, signle init, and hacked methods) https://www.vogons.org/viewtopic.php?t=52806. He wrote a nice seamless playback tool, too (tested this PR with it, and it's fine).

> I'm just cautious

Absolutely; given my track record with regression, there's probably a regression /somewhere/ with /some/ game! :laughing: and I've planned for that by committing one change at a time, so it can be isolated.  

But so far, it seems to be working quite well - and gets SimLife working, too. It's possible other games will be fixed.



I know there some efforts to disassemble a couple of the SB DSPs out there -

Reply:
Thanks for the explanation, you've certainly done your homework then 😅 📓 🖊️  Would it be worthwhile to include some pointers about all that in a few sentences? At least a link to the Vogons thread would be very helpful for the next person deep-investigating the SB DMA situation.

Reply:
Will add some of that, and add most of it to the commit comment block (so it's easy to "git log -S")!

Reply:
Ok - added!